File: JavaScript/543-Diameter of Binary Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var diameterOfBinaryTree = function(root) {
    if (!root) return 0;
    let res = 0;
    
    let dfs = (root)=>{
        if (!root) return -1;
        const left = dfs(root.left);
        const right = dfs(root.right);
        res = Math.max(res, left+right+2);
        return Math.max(left, right) + 1;
    }
    
    dfs(root);
    return res;
};

File: C++/543-Diameter of Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int res = 0;
    int dfs(TreeNode* node) {
        if (!node) return -1;
        int l = 1 + dfs(node->left);
        int r = 1 + dfs(node->right);
        res = max(res, l+r);
        return max(l, r);
    }
    int diameterOfBinaryTree(TreeNode* root) {
        dfs(root);
        return res;
    }
};